core/drivers/regulator/regulator_fixed.c

```diff
@@ -0,0 +1,186 @@
+// SPDX-License-Identifier: BSD-2-Clause
+/*
+ * Copyright (c) 2022-2023, STMicroelectronics
+ */
+
+#include <compiler.h>
+#include <drivers/gpio.h>
+#include <drivers/regulator.h>
+#include <dt-bindings/gpio/gpio.h>
+#include <kernel/delay.h>
+#include <kernel/dt_driver.h>
+#include <libfdt.h>
+#include <tee_api_types.h>
+#include <trace.h>
+
+static_assert(GPIO_LEVEL_HIGH == 1 && GPIO_LEVEL_LOW == 0);
+
+/*
+ * struct regulator_gpio - GPIO controlled regulator
+ * @regulator: Preallocated regulator instance
+ * @enable_gpio: GPIO for the enable state of the regulator or NULL if always on
+ * @enable_delay: Time (in microsecond) for the regulator to get enabled
+ * @off_on_delay: Min time (in microsecond) between enable and disable request
+ * @off_on_us: Timestamp of the last disable request
+ */
+struct regulator_fixed {
+	struct regulator regulator;
+	struct gpio *enable_gpio;
+	unsigned int enable_delay;
+	unsigned int off_on_delay;
+	uint64_t off_on_us;
+};
+
+static struct regulator_fixed *regulator_priv(struct regulator *regulator)
+{
+	return container_of(regulator, struct regulator_fixed, regulator);
+}
+
+static TEE_Result fixed_set_state(struct regulator *regulator, bool enabled)
+{
+	struct regulator_fixed *regu = regulator_priv(regulator);
+
+	if (regu->enable_gpio) {
+		if (enabled) {
+			while (!timeout_elapsed(regu->off_on_us))
+				udelay(1);
```

Comment:
do you think it's better to wait 1 microsecond here or to actively wait until timeout is elapsed?

read_time() API is based on ARM counter so maybe we can save some time by doing that but I guess the CPU will not be relaxed. I'm saying that because of how much time we might spend if the platform is stressed by loads of SCMI requests or other kind of calls.

Reply:
Indeed, the call to `udelay()` could be removed.
I would prefer it remains so it can easier evolve to something else (like a CPU relaxed polling), some day.